internal/controller/clusterorder_controller.go

```diff
@@ -19,47 +19,261 @@ package controller
+		{
+			NamespacedName: client.ObjectKey{
+				Name:      clusterOrderName,
+				Namespace: clusterOrderNamespace,
+			},
+		},
+	}
+}
+
+func namespaceNameFromInstance(instance *cloudkitv1alpha1.ClusterOrder) string {
+	return fmt.Sprintf("cluster-%s-%s", instance.GetName(), rand.String(6))
+}
+
+func (r *ClusterOrderReconciler) handleUpdate(ctx context.Context, _ ctrl.Request, instance *cloudkitv1alpha1.ClusterOrder) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Create or update " + instance.GetName())
+
+	// Do we have a finalizer yet?
+	if !controllerutil.ContainsFinalizer(instance, cloudkitFinalizer) {
+		controllerutil.AddFinalizer(instance, cloudkitFinalizer)
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Do we have a clusterReference yet?
+	if instance.Status.ClusterReference == nil {
+		log.Info("Adding cluster reference")
+		nsName := namespaceNameFromInstance(instance)
+		instance.Status.ClusterReference = newClusterReference(nsName)
+		if err := r.Status().Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	for _, component := range r.components() {
+		log.Info("Handling component " + component.name)
+
+		resource, err := component.fn(ctx, instance)
+		if err != nil {
+			log.Error(err, "Failed to mutate resource", component.name)
+			return ctrl.Result{}, err
+		}
+
+		if resource.shouldExist {
+			result, err := controllerutil.CreateOrUpdate(ctx, r.Client, resource.object, resource.mutateFn)
+			if err != nil {
+				log.Error(err, "Failed to create or update "+component.name)
+				return ctrl.Result{}, err
+			}
+			switch result {
+			case controllerutil.OperationResultCreated:
+				log.Info("Created " + component.name)
+			case controllerutil.OperationResultUpdated:
+				log.Info("Updated " + component.name)
+			}
+			continue
+		}
+
+		// If we get this far, resource should not exist
+		// Ensure the resource does not exist, and call Delete if necessary
+		key := client.ObjectKeyFromObject(resource.object)
+		if err := r.Client.Get(ctx, key, resource.object); err != nil {
+			if apierrors.IsNotFound(err) {
+				// "not found" is the desired state. Nothing else to do.
+				continue
+			}
+			log.Error(err, "Get request for resource failed", component.name)
+			return ctrl.Result{}, err
+		}
+		err = r.Client.Delete(ctx, resource.object)
+		if err != nil {
+			log.Error(err, "Delete request for resource failed", component.name)
+			return ctrl.Result{}, err
+		}
+		log.Info("Deleted " + component.name)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClusterOrderReconciler) handleDelete(ctx context.Context, _ ctrl.Request, instance *cloudkitv1alpha1.ClusterOrder) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Delete " + instance.GetName())
+
+	if controllerutil.ContainsFinalizer(instance, cloudkitFinalizer) {
+		// Can't do anything without status.clusterReference
+		if instance.Status.ClusterReference == nil {
```

Comment:
Along with my above comment, this should find the namespace ideally by label. Labels are part of a resource's identity, and it's the go-to way to associate resources when an owner reference isn't possible.

A guiding principle of k8s APIs is that you should be able to delete the status of a resource at any time, and the controller will reconstruct it from fresh observations.

Also, the status is just a point-in-time observation from the last time the controller was able to look. It makes no guarantees about the current state of the world.